pkg/datapath/iptables/reconciler.go

```diff
@@ -0,0 +1,203 @@
+		return false
+	}
+	return true
+}
+
+func toLocalNodeInfo(n node.LocalNode) localNodeInfo {
+	var (
+		v4AllocCIDR, v6AllocCIDR                 string
+		v4NativeRoutingCIDR, v6NativeRoutingCIDR string
+	)
+
+	if n.IPv4AllocCIDR != nil {
+		v4AllocCIDR = n.IPv4AllocCIDR.String()
+	}
+	if n.IPv6AllocCIDR != nil {
+		v6AllocCIDR = n.IPv6AllocCIDR.String()
+	}
+	if n.IPv4NativeRoutingCIDR != nil {
+		v4NativeRoutingCIDR = n.IPv4NativeRoutingCIDR.String()
+	}
+	if n.IPv6NativeRoutingCIDR != nil {
+		v6NativeRoutingCIDR = n.IPv6NativeRoutingCIDR.String()
+	}
+
+	return localNodeInfo{
+		internalIPv4:          n.GetCiliumInternalIP(false),
+		internalIPv6:          n.GetCiliumInternalIP(true),
+		ipv4AllocCIDR:         v4AllocCIDR,
+		ipv6AllocCIDR:         v6AllocCIDR,
+		ipv4NativeRoutingCIDR: v4NativeRoutingCIDR,
+		ipv6NativeRoutingCIDR: v6NativeRoutingCIDR,
+	}
+}
+
+type proxyInfo struct {
+	name        string
+	port        uint16
+	isIngress   bool
+	isLocalOnly bool
+}
+
+type noTrackPodInfo struct {
+	ip   netip.Addr
+	port uint16
+}
+
+func reconciliationLoop(
+	ctx context.Context,
+	log logrus.FieldLogger,
+	health cell.HealthReporter,
+	installIptRules bool,
+	params *reconcilerParams,
+	updateRules func(state desiredState, firstInit bool) error,
+) error {
+	// The minimum interval between reconciliation attempts
+	const minReconciliationInterval = time.Second / 5
+
+	state := desiredState{
+		installRules: installIptRules,
+		proxies:      sets.New[proxyInfo](),
+		noTrackPods:  sets.New[noTrackPodInfo](),
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	localNodeEvents := stream.ToChannel(ctx, params.localNodeStore)
+	state.localNodeInfo = toLocalNodeInfo(<-localNodeEvents)
+
+	devices, devicesWatch := tables.SelectedDevices(params.devices, params.db.ReadTxn())
+	state.devices = sets.New(tables.DeviceNames(devices)...)
+
+	// Use a ticker to limit how often the desired state is reconciled to avoid doing
+	// lots of operations when e.g. ipset updates.
+	ticker := time.NewTicker(minReconciliationInterval)
+	defer ticker.Stop()
+
+	// stateChanged is true when the desired state has changed or when reconciling it
+	// has failed. It's set to false when reconciling succeeds.
+	stateChanged := true
+
+	firstInit := true
+
+stop:
+	for {
+		select {
+		case <-ctx.Done():
+			break stop
+		case <-devicesWatch:
+			devices, devicesWatch = tables.SelectedDevices(params.devices, params.db.ReadTxn())
+			newDevices := sets.New(tables.DeviceNames(devices)...)
+			if newDevices.Equal(state.devices) {
+				continue
+			}
+			state.devices = newDevices
+			stateChanged = true
+		case localNode, ok := <-localNodeEvents:
+			if !ok {
+				break stop
+			}
+			localNodeInfo := toLocalNodeInfo(localNode)
+			if localNodeInfo.equal(state.localNodeInfo) {
+				continue
+			}
+			state.localNodeInfo = localNodeInfo
+			stateChanged = true
+		case proxyInfo, ok := <-params.proxies:
```

Comment:
Shouldn't these channels be drained as otherwise the methods sending to these could block if the order is unfortunate when stopping?